Licensing: update topic (#5815)
DevExpress/devextreme-documentation: concepts/Common/Licensing/10 How to Obtain and Validate Your License Key.md

    <!--JavaScript-->import config from 'devextreme/core/config'; 
    import { licenseKey } from './devextreme-license'; 
    
    config({ licenseKey });   

If your project does not use bundlers and includes sources via script tags, you can add a reference to a file holding the license key registration: 

    <!--JavaScript--><script src="./dx.all.js" type="text/javascript"> </script> 
    <script src="./devextreme-key.js" type="text/javascript"></script>

    <!-- tab: devextreme-key.js --> DevExpress.config({ licenseKey: '' });

### Automated License Key File Creation

Instead of manually creating the `devextreme-license.ts` file to store the license key, you can use a script that will automatically generate the file. The following code is an example of what such a script might look like:

    <!--JavaScript-->const fs = require('fs') 
    const path = './src/devextreme-license.ts'; 
    const key = process.env.DEVEXTREME_KEY ?? ''; 

    if (key || !fs.existsSync(path)) { 
        fs.writeFileSync(path, `export const licenseKey = '${key}';`); 
    }

> As shown above, if the `src/devextreme-license.ts` file is absent, the script creates it and assigns the value of the `DEVEXTREME_KEY` environment variable as the key enable value. Utilizing the `DEVEXTREME_KEY` variable is beneficial in CI; however, if you wish to use it locally, ensure that you initialize it with the license key value before executing the script.

Reviewer: if `DEVEXTREME_KEY` is set we use it to assign to the key, otherwise the key is an empty string. Should we highlight it or everybody can understand this piece of code?

You can call the `node add-devextreme-license` command to invoke the script manually, but it proves more useful to include its invocation in npm's `postinstall` script:

    <!-- tab: package.json -->{ 
        "scripts": { 
            "postinstall": "node add-devextreme-license" 
        }, 
    }

In this case, upon installing the NPM modules, each developer will receive an automatically created file to store the license key.

> Next, enter the key within [GlobalConfig](/api-reference/50%20Common/Object%20Structures/GlobalConfig '/Documentation/ApiReference/Common/Object_Structures/GlobalConfig') as described in the previous section.

Reviewer: in this approach the `src/devextreme-license.ts` still have to be included in the .gitignore file

### CI Licensing

> Use the `add-devextreme-license.js` script (as described in the previous section) to organize a license key check on CI. Set the `DEVEXTREME_KEY` environment variable on the remote machine.

Reviewer: It is not mandatory to use the script, it's just an option that we recommend.
I think it's better to say: You can use the  `add-devextreme-license.js` in the post install action to organize ...
In this case, you will need to set the `DEVEXTREME_KEY` environment variable on the remote machine.

    <!-- tab: devextreme-license.ts -->export const licenseKey = 'DEVELOPER_LICENSE_KEY’;

In order for each developer to use their own license key, the file should not be stored in the source code repository. Therefore, it is important to instruct Git to ignore the file that holds the key. To do this, you need to add the file path to your project's `.gitignore` file:

    <!-- tab: .gitignore -->src/devextreme-key.ts

This action also prevents your team from accidentally committing the key.

DevExtreme license key must be entered within [GlobalConfig](/api-reference/50%20Common/Object%20Structures/GlobalConfig '/Documentation/ApiReference/Common/Object_Structures/GlobalConfig'). This should be done in the entry point of the application:

    <!--JavaScript-->import config from 'devextreme/core/config'; 
    import { licenseKey } from './devextreme-license'; 
    
    config({ licenseKey });   

If your project does not use bundlers and includes sources via script tags, you can add a reference to a file holding the license key registration: 

>     <!--JavaScript--><script src="./dx.all.js" type="text/javascript"> </script> 
>     <script src="./devextreme-key.js" type="text/javascript"></script>
> 
>     <!-- tab: devextreme-key.js --> DevExpress.config({ licenseKey: '' });

Reviewer 2: ```suggestion
    <!--JavaScript--><script src="./dx.all.js" type="text/javascript"> </script> 
    <script src="./devextreme-license.js" type="text/javascript"></script>

    <!-- tab: devextreme-license.js --> DevExpress.config({ licenseKey: '' });
```

    <!--JavaScript-->const fs = require('fs') 
    const path = './src/devextreme-license.ts'; 
    const key = process.env.DEVEXTREME_KEY ?? ''; 

    if (key || !fs.existsSync(path)) { 
        fs.writeFileSync(path, `export const licenseKey = '${key}';`); 
    }

As shown above, if the `src/devextreme-license.ts` file is absent, the script creates it and assigns the value of the `DEVEXTREME_KEY` environment variable as the key enable value. Utilizing the `DEVEXTREME_KEY` variable is beneficial in CI; however, if you wish to use it locally, ensure that you initialize it with the license key value before executing the script.

[note] If you do not specify the `DEVEXTREME_KEY`, the key will be empty.

You can call the `node add-devextreme-license` command to invoke the script manually, but it proves more useful to include its invocation in npm's `postinstall` script:

    <!-- tab: package.json -->{ 
        "scripts": { 
            "postinstall": "node add-devextreme-license" 
        }, 
    }

> In this case, upon installing the NPM modules, each developer will receive an automatically created file to store the license key.

Reviewer 2: ```suggestion
In this case, after installing the NPM modules, each developer will receive an automatically created file to store the license key.
```

> Next, enter the key within [GlobalConfig](/api-reference/50%20Common/Object%20Structures/GlobalConfig '/Documentation/ApiReference/Common/Object_Structures/GlobalConfig') add the `devextreme-license.ts` file path to your project's `.gitignore` file as described in the previous section.

Reviewer 2: ```suggestion
Next, register the key within [GlobalConfig](/api-reference/50%20Common/Object%20Structures/GlobalConfig '/Documentation/ApiReference/Common/Object_Structures/GlobalConfig') add the `devextreme-license.ts` file path to your project's `.gitignore` file as described in the previous section.
```

### CI Licensing

You can use the `add-devextreme-license.js` script (as described in the previous section) to organize a license key check on CI. In this case, specify the `DEVEXTREME_KEY` environment variable on the remote machine.

In this topic, we consider the use of GitHub Actions and GitHub Secrets. A similar approach can be adopted for any other CI server.

> 1. Install GitHub repository-level secrets according to the [GitHub guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository).

Reviewer 2: ```suggestion
1. Set GitHub repository-level secret according to the [GitHub guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository).
```

To obtain a license key, you must navigate to the DevExpress [Download Manager](https://www.devexpress.com/ClientCenter/DownloadManager/) and select DevExtreme Subscription from your licensed product list. Follow on-screen instructions to copy your key.

![DevExtreme Download Manager](/images/Common/licensing-key@2x.png)

When multiple developers work on the project, each of them has their own personal license key, which they must specify for every machine they work on.

> ### Manual Configuration (Simplest Case)

Reviewer 2: ```suggestion
### Manual Configuration
```

In order for each developer to use their own license key, the file should not be stored in the source code repository. Therefore, it is important to instruct Git to ignore the file that holds the key. To do this, you need to add the file path to your project's `.gitignore` file:

    <!-- tab: .gitignore -->src/devextreme-key.ts

This action also prevents your team from accidentally committing the key.

DevExtreme license key must be entered within [GlobalConfig](/api-reference/50%20Common/Object%20Structures/GlobalConfig '/Documentation/ApiReference/Common/Object_Structures/GlobalConfig'). This should be done in the entry point of the application:

    <!--JavaScript-->import config from 'devextreme/core/config'; 
    import { licenseKey } from './devextreme-license'; 
    
    config({ licenseKey });   

If your project does not use bundlers and includes sources via script tags, you can add a reference to a file holding the license key registration: 

    <!--JavaScript--><script src="./dx.all.js" type="text/javascript"> </script> 
    <script src="./devextreme-license.js" type="text/javascript"></script>

>     <!-- tab: devextreme-license.js --> DevExpress.config({ licenseKey: '' });

Reviewer 2: ```suggestion
    <!-- tab: devextreme-license.js --> DevExpress.config({ licenseKey: 'DEVELOPER_LICENSE_KEY' });
```

### Automated License Key File Creation

Instead of manually creating the `devextreme-license.ts` file to store the license key, you can use a script that will automatically generate the file. The following code is an example of what such a script might look like:

    <!--JavaScript-->const fs = require('fs') 
    const path = './src/devextreme-license.ts'; 
    const key = process.env.DEVEXTREME_KEY ?? ''; 

    if (key || !fs.existsSync(path)) { 
        fs.writeFileSync(path, `export const licenseKey = '${key}';`); 
    }

> As shown above, if the `src/devextreme-license.ts` file is absent, the script creates it and assigns the value of the `DEVEXTREME_KEY` environment variable as the key value. Utilizing the `DEVEXTREME_KEY` variable is beneficial in CI: ensure that you initialize the license key value before executing the script. If you use the script locally, you can add the license key in the file after it is generated.

Reviewer 2: ```suggestion
As shown above, if the `src/devextreme-license.ts` file is absent, the script creates it and assigns the value of the `DEVEXTREME_KEY` environment variable as the key value. Utilizing the `DEVEXTREME_KEY` variable is beneficial in CI. If you use the script locally, you can add the license key in the file after it is generated.
```

[note] If you do not specify the `DEVEXTREME_KEY`, the key will be empty.

You can call the `node add-devextreme-license` command to invoke the script manually, but it proves more useful to include its invocation in npm's `postinstall` script:

    <!-- tab: package.json -->{ 
        "scripts": { 
            "postinstall": "node add-devextreme-license" 
        }, 
    }

In this case, after installing the NPM modules, each developer will receive an automatically created file to store the license key.

Next, register the key within [GlobalConfig](/api-reference/50%20Common/Object%20Structures/GlobalConfig '/Documentation/ApiReference/Common/Object_Structures/GlobalConfig') add the `devextreme-license.ts` file path to your project's `.gitignore` file as described in the previous section.

### CI Licensing

> You can use the `add-devextreme-license.js` script (as described in the previous section) to organize a license key check on CI. In this case, specify the `DEVEXTREME_KEY` environment variable on the remote machine.

Reviewer 2: ```suggestion
You can use the `add-devextreme-license.js` script (as described in the previous section) to organize a license key check on CI. In this case, ensure that you initializethe `DEVEXTREME_KEY` environment variable on the remote machine before the script is  executed

```

In this topic, we consider the use of GitHub Actions and GitHub Secrets. A similar approach can be adopted for any other CI server.

> 1. Specify GitHub repository-level secret according to the [GitHub guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository).

Reviewer 2: ```suggestion
1. Specify GitHub repository-level secret `DevExtremeKey` according to the [GitHub guide](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository).
```
